### PR TITLE
Implement state saving

### LINKED ISSUE #215: Store state restoring
We discussed that it will be nice to have some kind of callback, that calls before the store is destroyed and allow the lib user to save part of the state to SavedStateHandle.

### PR DESCRIPTION
**Pull request by sergei-lapin**
Utilize saved state handle in order to provide ability to persist store's state

Closes #215

### CODE REVIEW

**Review by opengamer29 — APPROVED** [review]

File: elmslie-samples/android-loader/src/main/java/vivid/money/elmslie/samples/android/loader/MainActivity.kt [hunk 1]
```diff
@@ -34,22 +38,20 @@ class MainActivity : AppCompatActivity(R.layout.activity_main), ElmRendererDeleg
...
+        when (effect) {
+            Effect.ShowError ->
+                Snackbar.make(findViewById(R.id.content), "Error!", Snackbar.LENGTH_SHORT).show()
+        }
 
-    @Suppress("UnusedPrivateMember")
-    private fun createStore(savedStateHandle: SavedStateHandle): Store<Event, Effect, State> =
-        storeFactory()
+    private fun createStore(handle: SavedStateHandle): Store<Event, Effect, State> =
+        storeFactory(initial = handle.storeStateBundle.getInt(HANDLE_KEY).takeUnless { it == 0 })
```
Comment (rcmkt) [inline elmslie-samples/android-loader/src/main/java/vivid/money/elmslie/samples/android/loader/MainActivity.kt:56]: Can we hide getting `storeStateBundle` under the hood? For instance, getting two params in lambda: (SavedStateHandle, Bundle) ?

File: elmslie-android/src/main/java/vivid/money/elmslie/android/ElmStoreLazy.kt [hunk 2]
```diff
@@ -71,36 +77,62 @@ internal fun <
...
     storeFactory: (SavedStateHandle) -> Store<Event, Effect, State>,
+    saveState: (State, Bundle) -> Unit,
 ) : ViewModel() {
 
     val store = storeFactory.invoke(savedStateHandle).toCachedStore().also { it.start() }
 
+    init {
+        savedStateHandle.setSavedStateProvider(StateBundleKey) {
+            val result = bundleOf()
+            saveState.invoke(store.states().value, result)
```
[the remark this answers is not in the document]
Comment (sergei-lapin) [reply elmslie-android/src/main/java/vivid/money/elmslie/android/ElmStoreLazy.kt]: dope, did

File: elmslie-samples/android-loader/src/main/java/vivid/money/elmslie/samples/android/loader/MainActivity.kt (hunk 1 above)
Comment (sergei-lapin) [reply elmslie-samples/android-loader/src/main/java/vivid/money/elmslie/samples/android/loader/MainActivity.kt]: we can, but imo we shouldn't  
that's going to result in misleading api  
i'll think on other ways around it this week

File: elmslie-android/src/main/java/vivid/money/elmslie/android/ElmStoreLazy.kt [hunk 3]
```diff
@@ -71,36 +77,59 @@ internal fun <
...
+        }
+    }
+
     override fun onCleared() {
         store.stop()
     }
+
+    companion object {
+
+        @PublishedApi internal const val StateBundleKey = "state_bundle"
```
Comment (rcmkt) [inline elmslie-android/src/main/java/vivid/money/elmslie/android/ElmStoreLazy.kt:111]: Maybe remove 'internal' then?)

File: elmslie-samples/android-loader/src/main/java/vivid/money/elmslie/samples/android/loader/MainActivity.kt (hunk 1 above)
Comment (sergei-lapin) [reply elmslie-samples/android-loader/src/main/java/vivid/money/elmslie/samples/android/loader/MainActivity.kt]: i didn't manage to find better solution than to wrap both handle and bundle into single entity and make factory lambda to be invoked in the receiver context of it.  

i tried providing unifying interface for these types with single method (get) enabling querying both of them at the same time but there is an issue of type resolving for complex bundle types and duplicating all bundle get* methods seem verbose

thus i ended up with the current solution

File: elmslie-android/src/main/java/vivid/money/elmslie/android/ElmStoreLazy.kt (hunk 3 above)
Comment (sergei-lapin) [reply elmslie-android/src/main/java/vivid/money/elmslie/android/ElmStoreLazy.kt]: it's private now

File: elmslie-samples/android-loader/src/main/java/vivid/money/elmslie/samples/android/loader/MainActivity.kt (hunk 1 above)
Comment (rcmkt) [reply elmslie-samples/android-loader/src/main/java/vivid/money/elmslie/samples/android/loader/MainActivity.kt]: I think it's a bit overengineered with the additional entity, can be simplified with the two params in the lambda)
I don't see the way of misleading api to be honest)

Comment (rcmkt) [reply elmslie-samples/android-loader/src/main/java/vivid/money/elmslie/samples/android/loader/MainActivity.kt]: @opengamer29 We need the third opinion on this, what do you think? Is there some misleading here?

Comment (opengamer29) [reply elmslie-samples/android-loader/src/main/java/vivid/money/elmslie/samples/android/loader/MainActivity.kt]: All this is about decreasing cognitivе load. Additional class SavedStateExtras doesn't help in it.

I also think that passing both SavedStateHandle and Bundle as lambdas is a little bit strange (you can extract Bundle from SavedStateHandle)

I prefer the option with only SavedStateHandle as a parameter. It's a common pattern in Android (and this module will be used only for Android) and it's not a huge issue that developers should keep in mind keys what they use.

Comment (sergei-lapin) [reply elmslie-samples/android-loader/src/main/java/vivid/money/elmslie/samples/android/loader/MainActivity.kt]: ok then, i rolled back to the initial solution with saved state handle being the only arg of store factory and added comments to store functions

Comment (sergei-lapin) [reply elmslie-samples/android-loader/src/main/java/vivid/money/elmslie/samples/android/loader/MainActivity.kt]: while doing this i couldn't help but wander: why google implemented such contradictory behavior within the same entity? 🤷🏻‍♂️ 

we get implicit unpacking of default args bundle to saved state handle but for the user's saved state that is not the case...

Comment (rcmkt) [reply elmslie-samples/android-loader/src/main/java/vivid/money/elmslie/samples/android/loader/MainActivity.kt]: so be it) for the last change, maybe add some extra value to `StateBundleKey` (e.g. package) to protect from collision? I think "state_bundle" is a super common value)

**Review by rcmkt — APPROVED** [review]

**Pull request merged**